### PR TITLE
NIFI-1118 Update SplitText Processor - add support for split size limits and header line markers.

### PR DESCRIPTION
**Pull request by jskora**
* Add "Maximum Fragment Size" property.  A new split file will be created if the next line to be added to the current split file exceeds this user-defined maximum file size.  In the case where an input line is longer than the fragment size, this line will be output in a separate split file that will exceed the maximum fragment size.
* Add "Header Line Marker Character" property.  Lines that begin with these user-defined character(s) will be considered header line(s) rather than a predetermined number of lines.  The existing property "Header Line Count" must be zero for this new property and behavior to be used.
* Deprecated the "Remove Trailing Newlines" property.
* Fixed conditional that incorrectly suppressed splits where the content line count equaled the header line count and did not remove empty splits from the session.
* Minor formatting cleanup.
* Exclude test files from RAT check in pom.xml.

### CODE REVIEW

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java [hunk 1]
```diff
@@ -63,9 +71,16 @@
 @SupportsBatching
 @Tags({"split", "text"})
 @InputRequirement(Requirement.INPUT_REQUIRED)
-@CapabilityDescription("Splits a text file into multiple smaller text files on line boundaries, each having up to a configured number of lines")
+//@CapabilityDescription("Splits a text file into multiple smaller text files on line boundaries, each having up to a configured number of lines")
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: The commented-out line should probably be removed :)

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java [hunk 2]
```diff
@@ -134,6 +168,28 @@ protected void init(final ProcessorInitializationContext context) {
...
+        ArrayList<ValidationResult> results = new ArrayList<>();
+
+        results.add(new ValidationResult.Builder()
+            .subject("Remove Trailing Newlines")
+            .valid(!validationContext.getProperty(REMOVE_TRAILING_NEWLINES).asBoolean())
+            .explanation("Property is deprecated; value must be set to false. Future releases may remove this Property.")
+            .build());
+
+        final boolean invalidState = (validationContext.getProperty(LINE_SPLIT_COUNT).asInteger() == 0
+                && validationContext.getProperty(FRAGMENT_MAX_SIZE).asDataSize(DataUnit.B) == null);
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: I think this line should be:
&& !validationContext.getProperty(FRAGMENT_MAX_SIZE).isSet() );

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java [hunk 3]
```diff
@@ -134,6 +168,28 @@ protected void init(final ProcessorInitializationContext context) {
     }
 
     @Override
+    protected Collection<ValidationResult> customValidate(ValidationContext validationContext) {
+        ArrayList<ValidationResult> results = new ArrayList<>();
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: Is there any reason we are declaring the type here as ArrayList, rather than List?

Comment (jskora) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: I'm fixing these.  Update will be pushed shortly.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java [hunk 4]
```diff
@@ -143,72 +199,82 @@ protected void init(final ProcessorInitializationContext context) {
...
         for (int i = 0; i < maxNumLines; i++) {
-            final long bytes = countBytesToSplitPoint(in, out, keepAllNewLines || (i != maxNumLines - 1));
+            final long bytes = countBytesToSplitPoint(in, out, totalBytes, maxByteCount);
+            totalBytes += bytes;
             if (bytes <= 0) {
                 return numLines;
             }
-
             numLines++;
+            if (totalBytes >= maxByteCount && numLines > maxNumLines) {
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: The logic here appears to be incorrect.
numLines is incremented for each iteration of the loop (unless we return before it is incremented).
This means that numLines <= i
The loop's condition indicates i < maxNumLines
So numLines <= i < maxNumLines
So it is always the case that numLines < maxNumLines, so this condition will never be satisfied because numLines will never be > maxNumLines

Now, looking through the code and doing a bit of testing, this does not appear to return an incorrect result, since countBytesToSplitPoint will handle the logic appropriately itself, but this should be fixed before it is merged.

Comment (markobean) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: Agreed. "&& numLines > maxNumLines" condition can be removed.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java [hunk 5]
```diff
@@ -143,72 +199,82 @@ protected void init(final ProcessorInitializationContext context) {
...
-                bytesRead--;    // we reset the stream by 1 byte so decrement the number of bytes read by 1
-                return includeLineDelimiter ? bytesRead : bytesRead - 1;
+            // Determine if \n follows \r; for both cases, end of line has been reached
+            if (nextByte == '\r') {
+                buffer.writeTo(out);
+                buffer.close();
+                in.mark(1);
+                final int lookAheadByte = in.read();
+                if (lookAheadByte == '\n') {
+                    out.write(lookAheadByte);
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: This will throw a NullPointerException if out == null

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java [hunk 6]
```diff
@@ -143,72 +199,82 @@ protected void init(final ProcessorInitializationContext context) {
...
 
-            // we didn't get a new line but if last byte was carriage return we've reached a new-line.
-            // so we roll back the last byte that we read and return
-            if (lastByte == '\r') {
-                in.reset();
-                bytesRead--;    // we reset the stream by 1 byte so decrement the number of bytes read by 1
-                return includeLineDelimiter ? bytesRead : bytesRead - 1;
+            // Determine if \n follows \r; for both cases, end of line has been reached
+            if (nextByte == '\r') {
+                buffer.writeTo(out);
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: This will throw a NullPointerException if out == null

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java [hunk 7]
```diff
@@ -143,72 +199,82 @@ protected void init(final ProcessorInitializationContext context) {
...
         }
-
         return numLines;
     }
 
-    private long countBytesToSplitPoint(final InputStream in, final OutputStream out, final boolean includeLineDelimiter) throws IOException {
-        int lastByte = -1;
+    private long countBytesToSplitPoint(final InputStream in, final OutputStream out, final long bytesReadSoFar, final long maxSize) throws IOException {
         long bytesRead = 0L;
+        final ByteArrayOutputStream buffer = new ByteArrayOutputStream();
```
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java]: It looks like we always buffer this data into Java heap, but we don't actually use it for anything if out == null (which is the case more often than not, I believe). This is pretty concerning, as we should avoid buffering potentially large amounts of data into Java heap unless absolutely necessary. I think this needs to be reworked so that we don't buffer the data here, if out == null, since we won't make use of it anyway.

**Comment by jskora** [comment]
I think the issues are all covered now and tests have been expanded to get all critical logic and edge cases.

**Comment by joewitt** [comment]
@jskora do you think this PR can be closed now given the updates made to fix the underlying defects found?  A new PR could be submitted which adds the proposed features or goes into ReplaceText or a new processor.

**Comment by markobean** [comment]
@joewitt Talked with @jskora and concur this PR can be closed. A new PR will be opened later with the new features and also with the Return Trailing Newlines bug fix included (if RTN is still included.)

**Comment by joewitt** [comment]
thanks @markobean and @jskora .  What do you think about making ignore newlines only be honored/supported when not using the new features you're planning to include or only in very specific configurations?  I ask because this, admittedly mistaken, feature is used a lot.  Ultimately if that seems to unwieldy we can punt that feature in 1.0, add your new capabilities, and support end of line removal on ReplaceText instead.  We just need to remember to document this in the migration guide for 1.0 as this could cause some pretty funky behavior changes for folks.

What do you think?

**Comment by mosermw** [comment]
@jskora Are we closing this PR?  It appears the latest PR for NIFI-1118 is [PR#444](https://github.com/apache/nifi/pull/444)

**Comment by jskora** [comment]
Yes, this should be closed.  Done.
